Packages/com.unity.inputsystem/InputSystem/Plugins/XR/Controls/PoseControl.cs

```diff
@@ -209,26 +211,66 @@ protected override void FinishSetup()
-                trackingState = (TrackingState)trackingState.ReadUnprocessedValueFromState(statePtr),
-                position = position.ReadUnprocessedValueFromState(statePtr),
-                rotation = rotation.ReadUnprocessedValueFromState(statePtr),
-                velocity = velocity.ReadUnprocessedValueFromState(statePtr),
-                angularVelocity = angularVelocity.ReadUnprocessedValueFromState(statePtr),
-            };
+                case OptimizedControlDataType.Pose:
+                    return *(PoseState*)((byte*)statePtr + (int)m_StateBlock.byteOffset);
+                default:
+                    return new PoseState()
+                    {
+                        isTracked = isTracked.ReadUnprocessedValueFromState(statePtr) > 0.5f,
+                        trackingState = (TrackingState)trackingState.ReadUnprocessedValueFromState(statePtr),
+                        position = position.ReadUnprocessedValueFromState(statePtr),
+                        rotation = rotation.ReadUnprocessedValueFromState(statePtr),
+                        velocity = velocity.ReadUnprocessedValueFromState(statePtr),
+                        angularVelocity = angularVelocity.ReadUnprocessedValueFromState(statePtr),
+                    };
+            }
         }
 
         /// <inheritdoc />
         public override unsafe void WriteValueIntoState(PoseState value, void* statePtr)
         {
-            isTracked.WriteValueIntoState(value.isTracked, statePtr);
-            trackingState.WriteValueIntoState((uint)value.trackingState, statePtr);
-            position.WriteValueIntoState(value.position, statePtr);
-            rotation.WriteValueIntoState(value.rotation, statePtr);
-            velocity.WriteValueIntoState(value.velocity, statePtr);
-            angularVelocity.WriteValueIntoState(value.angularVelocity, statePtr);
+            EnsureOptimizationTypeDoesntSuddenlyChange();
+
+            switch (m_OptimizedControlDataType)
+            {
+                case OptimizedControlDataType.Pose:
+                    *(PoseState*)((byte*)statePtr + (int)m_StateBlock.byteOffset) = value;
+                    break;
+                default:
+                    isTracked.WriteValueIntoState(value.isTracked, statePtr);
+                    trackingState.WriteValueIntoState((uint)value.trackingState, statePtr);
+                    position.WriteValueIntoState(value.position, statePtr);
+                    rotation.WriteValueIntoState(value.rotation, statePtr);
+                    velocity.WriteValueIntoState(value.velocity, statePtr);
+                    angularVelocity.WriteValueIntoState(value.angularVelocity, statePtr);
+                    break;
+            }
+        }
+        
+        public OptimizedControlDataType CalculateOptimizedControlDataType()
+        {
+            if (
+                m_StateBlock.sizeInBits == 60 * 8 &&
```

Comment:
Should be PoseState.kSizeInBytes instead of 60?

Reply:
Will fix but it doesn't matter that much, as you can see down in the check we also check layout of controls, so if anything about `PoseState` changes we also need to change that code as well.

Reply:
Fixed